File: exercise4.cpp
/*
4. Create an online shipping system with the following features:
* 		The system has a login that locks after the third failed attempt.
* 		Display a menu that allows: Sending a package, exiting the system.
* 		To send a package, sender and recipient details are required.
* 		The system assigns a random package number to each sent package.
* 		The system calculates the shipping price. $2 per kg.
* 		The user must input the total weight of their package, and the system should display the amount to pay.
* 		The system should ask if the user wants to perform another operation. If the answer is yes, it should return to the main menu. If it's no, it should close the system.

*/

#include <iostream>
#include <string>
using namespace std;

const string fix_user = "admin";
const string fix_pass = "1234";

string user = "";
string pass = "";
string sender = "";
string addressee = "";
string city = "";

char deliver = ' ';

int tries = 3, menu = 0, weight = 0;

bool access_flag = false, repeat_flag = false;

int main()
{

    cout << "Welcome to the Shipping System\n";
    do
    {
        cout << "Please put your user and password.\n";
        cout << "User: \n";
        cin >> user;
        cout << "Password: \n";
        cin >> pass;
        if (user == fix_user && pass == fix_pass)
        {
            cout << "Access granted \n";
            access_flag = true;
        }
        else
        {
            tries--;
            cout << "Wrong Data, " << tries << " left.\n";
            if (tries == 0)
            {
                cout << "Closing Shipping System...";
                exit(0);
            }
        }
    } while (access_flag = false);

    do
    {
        repeat_flag = false;
        cout << "Main Menu: \n";
        cout << "1. Send a Package: \n";
        cout << "2. Exit: \n";
        cout << "Please select one option: ";
        cin >> menu;

        switch (menu)
        {
        case 1:
            cout << "Please specify the sender of the package: \n";
            cin >> sender;
            cout << "Please specify the addressee of the package: \n";
            cin >> addressee;
            cout << "Please specify the city of delivery of the package: \n";
            cin >> city;
            cout << "Please specify the total weight (in Kg) of the package: \n";
            cin >> weight;
            cout << "Your delivery will cost: $" << (weight * 2) << " USD \n";
            cout << "Do you agree? (Y/N): ";
            cin >> deliver;
            if (deliver == 'Y' || deliver == 'y')
            {
                cout << "Thanks! Your shipping details are this: \n";
                cout << "Shippin number: " << (rand()) << endl;
                cout << "Sender: " << sender << endl;
                cout << "Addressee: " << addressee << endl;
                cout << "City: " << city << endl;
                cout << "Weight: " << weight << endl;
                cout << "Final Cost: $" << (weight * 2) << "USD \n";
                cout << " \n";
                cout << "Do you want to perform another operation? \n";
                cin >> deliver;
                if (deliver == 'Y' || deliver == 'y')
                {
                    repeat_flag = true;
                }
                else
                {
                    cout << "Thanks for using our services.\n";
                    cout << "Closing Shipping System...";
                    exit(0);
                }
            }
            else
            {
                repeat_flag = true;
            }
            break;

        case 2:
            cout << "Closing Shipping System...";
            exit(0);
            break;

        default:
            cout << "Please select a valid option \n";
            repeat_flag = true;
            break;
        }
    } while (repeat_flag = true);
}

File: Exercise4.py
#Exercise 4#
#!Create an online shipping system with the following features:
#|The system has a login that locks after the third failed attempt.
#|Display a menu that allows: Sending a package, exiting the system.
#|To send a package, sender and recipient details are required.
#|The system assigns a random package number to each sent package.
#|The system calculates the shipping price. $2 per kg.
#|The user must input the total weight of their package, and the system should display the amount to pay.
#|The system should ask if the user wants to perform another operation. If the answer is yes, it should return to the main menu. If it's no, it should close the system.

import pycountry
import random
import string

users = {
    "Mike": "123",
    "Peter": "123",
    "Tony": "123",
    "1":"1"
}

def shipping():
    def repeat():
        selection = input("Would you like to perform another operation? (Y/N): ")
        match selection.upper():
            case "Y":
                return True
            case "N":
                return False
            case _:
                print("Invalid input.")
                return repeat()

    def random_package_number():
        characters = string.ascii_uppercase + string.digits
        tracking_number = ''.join(random.choice(characters) for i in range(8))
        return tracking_number
    
    def package_weight():
        weight = input("Insert the weight (in kg) of your package: ")
        print("=" * 100)
        if(weight.isnumeric()):
            return int(weight)
        else:
            print("Invalid input")
            return package_weight()

    while True:
        print("=" * 100)
        sender_details_country = input("To which country you're sending the package? ")
        if(pycountry.countries.get(name=sender_details_country)):
            print("=" * 100)
            print(f"Your total for the shipping is: ${package_weight() * 2} dollars")
            print(f"The tracking number of your package is: {random_package_number()}")

            if(repeat()):
                main()
                break
            else:
                break
        else:
            print("The country doesnt exists or is spelled incorrectly. Try Again")

def main():
    while True:
        print("=" * 100)
        print("Welcome to the online shipping system\n1.Send a package\n2.Exit the system")
        selection = input("What action would you like to do?: ")
        if selection.isnumeric():
            match selection:
                case "1":
                    shipping()
                    print("Thank you for using the system!")
                    break
                case "2":
                    print("Thank you for using the system!")
                    break
        else:
            print("Invalid input. Try again")

def validate_user(username, password):
    if(users[username] == password):
        return True
    else:
        return False

def login():
    attempts = 0
    while(attempts < 3):
        print("=================================")
        user_username = input("Enter your username: ")
        if(user_username in users):
            user_password = input("Enter your password: ")
            if(validate_user(user_username, user_password)):
                print("*Login succesfull*")
                main()
                break
            else:
                print("Incorrect password. Try again.")
                attempts += 1
        else:
            print("Username does not exist. Try again.")
    if(attempts == 3):
        print("Too many unsuccessful attempts. The program is now locked.")

login()